core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -294,33 +298,74 @@ class BlockManagerMasterEndpoint(
       }
     }.toSeq
 
-    val removeRddBlockViaExtShuffleServiceFutures = externalBlockStoreClient.map { shuffleClient =>
-      blocksToDeleteByShuffleService.map { case (bmId, blockIds) =>
-        Future[Int] {
-          val numRemovedBlocks = shuffleClient.removeBlocks(
-            bmId.host,
-            bmId.port,
-            bmId.executorId,
-            blockIds.map(_.toString).toArray)
-          numRemovedBlocks.get(defaultRpcTimeout.duration.toSeconds, TimeUnit.SECONDS)
+    val removeRddBlockViaExtShuffleServiceFutures = if (externalShuffleServiceRddFetchEnabled) {
+      externalBlockStoreClient.map { shuffleClient =>
+        blocksToDeleteByShuffleService.map { case (bmId, blockIds) =>
+          Future[Int] {
+            val numRemovedBlocks = shuffleClient.removeBlocks(
+              bmId.host,
+              bmId.port,
+              bmId.executorId,
+              blockIds.map(_.toString).toArray)
+            numRemovedBlocks.get(defaultRpcTimeout.duration.toSeconds, TimeUnit.SECONDS)
+          }
         }
-      }
-    }.getOrElse(Seq.empty)
+      }.getOrElse(Seq.empty)
+    } else {
+      Seq.empty
+    }
 
     Future.sequence(removeRddFromExecutorsFutures ++ removeRddBlockViaExtShuffleServiceFutures)
   }
 
   private def removeShuffle(shuffleId: Int): Future[Seq[Boolean]] = {
-    // Nothing to do in the BlockManagerMasterEndpoint data structures
     val removeMsg = RemoveShuffle(shuffleId)
-    Future.sequence(
-      blockManagerInfo.values.map { bm =>
-        bm.storageEndpoint.ask[Boolean](removeMsg).recover {
-          // use false as default value means no shuffle data were removed
-          handleBlockRemovalFailure("shuffle", shuffleId.toString, bm.blockManagerId, false)
+    val removeShuffleFromExecutorsFutures = blockManagerInfo.values.map { bm =>
+      bm.storageEndpoint.ask[Boolean](removeMsg).recover {
+        // use false as default value means no shuffle data were removed
+        handleBlockRemovalFailure("shuffle", shuffleId.toString, bm.blockManagerId, false)
+      }
+    }.toSeq
+
+    // Find all shuffle blocks on executors that are no longer running
```

Comment:
Shall we unify these two cases together? e.g.,
```scala
mapOutputTracker.shuffleStatuses.get(shuffleId).map { shuffleStatus =>
  val (mapStatusAtExecutor, mapStatusAtShuffleService) = {
    shuffleStatus.mapStatuses.partition { mapStatus =>
      blockManagerIdByExecutor.contains(mapStatus.location.executorId)
    } match {
      case (mapStatusAtExecutor, _) if !externalShuffleServiceRemoveShuffleEnabled =>
        (mapStatusAtExecutor, Array.empty)
      case (se, ss) =>
        (se, ss)
    }
  }
  mapStatusAtExecutor.map(_.location).toSet.map {
   ...
  } ++ mapStatusAtShuffleService.map {
   ...
  }
}
```
This way makes us only send shuffle remove request to valid executors (not all active executors) in the first case.

Reply:
Fine. We could do it in a follow-up PR.

> And I'm not sure if all block managers still track shuffles for some reason even if they don't have any of the shuffle files?

If so, you should also send request to all the external shuffle service..

Reply:
The shuffle service doesn't tracking anything about shuffles, which is why we need to send it each individual block to delete instead of just a `RemoveShuffle` message